File: search/munchausen.c
// calculate munchausen numbers
#include <stdio.h>
#include <math.h>

int limit = 5000; // the upper bound of the search

int i;
int main() {
	for (i = 1; i < limit; i++) {
		// loop through each digit in i
		// e.g. for 1000 we get 0, 0, 0, 1.
		int number = i;
		int sum = 0;
		while (number > 0) {
			int digit = number % 10;
			number = number / 10;
			// find the sum of the digits 
			// raised to themselves 
			sum += pow(digit, digit);
		}
		if (sum == i) {
			// the sum is equal to the number
			// itself; thus it is a 
			// munchausen number
			printf("%i (munchausen)\n", i);
		} 
	}	
	return 0;
}

File: search/munchausen.clj
#!/usr/bin/env clojure
(defn ** [x n]
  (. (. java.math.BigInteger (valueOf x)) (pow n)))

(defn raise-to-itself [number]
  (** number number))

(defn digits [number]
  ; convert the number to a string,
  ; and convert each char to an int.
  ;
  ; subtract 48 because casting a char
  ; to an int returns the ASCII
  ; representation of that char.
  (map #(- (int %) 48) (str number))) 

(defn munchausen? [number]
  ; if the sum of the digits raised to
  ; themselves is equal to the number 
  ; itself, then it is a munchausen number.
  (= (apply + (map raise-to-itself (digits number))) number))

(def munchausen (filter munchausen? (range 5000)))
(println munchausen)
